### ISSUE TITLE
install_programo.php fails to start

### ISSUE THREAD

**Issue by pokephreak** [Program-O/Program-O#188]
I have Ubuntu 14.04 64 bit with webmin/virtualmin and LAMP stack, set up for program-o so if you know of a better config, please suggest.

When I go to install_programo.php the page seems to render the php as text instead of running it.

Initially I got this warning in /var/log/apache2/error.log : 

`PHP Fatal error:  require_once(): Failed opening required '../config/global_config.php' (include_path='.:/usr/share/php:/usr/share/pear') in /var/www/html/Program-O/install/index.php on line 15`

but this went away after I added /var/www/html to the php .ini files. I still see the same behavior in the browser at my.server.ip/Program-O/install/install_programo.php

I have tried all sorts of mods to server files over the past couple of days trying to fix this and can't get it running. I have tried both branches of the repo. I can confirm php is working on the server.

I have verified the required permissions are applied to the files mentioned in the install script. Here is a full paste of the page I see when I go to install_programo.php

`We're sorry, but Program O requires PHP version 5.3.0 or greater to function. Please ask your hosting provider to upgrade.

"; If (!$pdoSupport) $fatalError .= "

Support for PHP Data Objects (PDO) was not detected! This is required for Program O to function. Please ask your hosting provider to upgrade.
"; $no_unicode_message = (function_exists('mb_check_encoding')) ? '' : "

Warning! Unicode Support is not available on this server. Non-English languages will not display properly. Please ask your hosting provider to enable the PHP mbstring extension to correct this.
\n"; $errorMessage = (!empty ($_SESSION['errorMessage'])) ? $_SESSION['errorMessage'] : ''; $errorMessage .= $no_unicode_message; require_once ('install_config.php'); $dirArray = glob(_ADMIN_PATH_ . "ses_*",GLOB_ONLYDIR); $session_dir = (empty($dirArray)) ? 'ses_' . md5(time()) : basename($dirArray[0]); $dupPS = "$path_separator$path_separator"; $session_dir = str_replace($dupPS, $path_separator, $session_dir); // remove double path separators when necessary define('_SESSION_PATH_', _ADMIN_PATH_ . $session_dir . DIRECTORY_SEPARATOR); $writeCheckArray = array('config' => _CONF_PATH_, 'debug' => _DEBUG_PATH_, 'logs' => _LOG_PATH_); $errFlag = false; foreach ($writeCheckArray as $key => $folder) { if (!is_writable($folder)) { $errFlag = true; error_log("The folder $folder is not writable.", 3, '../logs/install.log'); $errorMessage .= << The $key folder cannot be written to, or does not exist. Please correct this before you continue.

endWarning; } } $additionalInfo = << This is usually a permissions issue, and most often occurs with Linux-based systems. Check file/folder permissions for the following directories:

    The base install folder (where you unzipped or uploaded the script to)
    admin
    config
    chatbot/debug

Permissions for these folders should be 0755. If they are not, then you need to change that. If you have trouble with this, or have questions, please visit us at Program O.

endInfo; if ($errFlag) $errorMessage .= $additionalInfo; define('SECTION_START', ''); # search params for start and end of sections define('SECTION_END', ''); # search params for start and end of sections define('PHP_SELF', $_SERVER['SCRIPT_NAME']); # This is more secure than $_SERVER['PHP_SELF'], and returns more or less the same thing ini_set("display_errors", 0); ini_set("log_errors", true); ini_set("error_log", _LOG_PATH_ . "install.error.log"); $myHost = $_SERVER['SERVER_NAME']; chdir(dirname(realpath(__FILE__))); $page_template = file_get_contents('install.tpl.htm'); $page = (isset ($_REQUEST['page'])) ? $_REQUEST['page'] : 1; $action = (isset ($_REQUEST['action'])) ? $_REQUEST['action'] : ''; $message = ''; if (!empty ($action)) { $message = $action($page); } $pageTemplate = 'Container'; $pageNotes = ucwords("Page $page Notes"); $content = getSection('Header', $page_template, false); $content .= getSection($pageTemplate, $page_template); $content .= getSection('Footer', $page_template); $content .= getSection("jQuery$page", $page_template); $notes = getSection($pageNotes, $page_template); $submitButton = getSection('SubmitButton', $page_template); $main = ($page == 1) ? getSection('InstallForm', $page_template) : $message; $main = (empty($fatalError)) ? $main : $fatalError; $tmpSearchArray = array(); $content = str_replace('[mainPanel]', $main, $content); $content = str_replace('[http_host]', $myHost, $content); $content = str_replace('[error_response]', $error_response, $content); $content = str_replace('[notes]', $notes, $content); $content = str_replace('[PHP_SELF]', PHP_SELF, $content); $content = str_replace('[errorMessage]', $errorMessage, $content); $content = str_replace('[cr6]', "\n ", $content); $content = str_replace('[cr4]', "\n ", $content); $content = str_replace("\r\n", "\n", $content); $content = str_replace("\n\n", "\n", $content); $content = str_replace('[admin_url]', _ADMIN_URL_, $content); $content .= << endPage; exit($content); /** * Function getSection * * * @param $sectionName * @param $page_template * @param bool $notFoundReturn * @return string */ function getSection($sectionName, $page_template, $notFoundReturn = true) { $sectionStart = str_replace('[section]', $sectionName, SECTION_START); $sectionStartLen = strlen($sectionStart); $sectionEnd = str_replace('[section]', $sectionName, SECTION_END); $startPos = strpos($page_template, $sectionStart, 0); if ($startPos === false) { if ($notFoundReturn) { return ''; } else $startPos = 0; } else $startPos += $sectionStartLen; $endPos = strpos($page_template, $sectionEnd, $startPos) - 1; $sectionLen = $endPos - $startPos; $out = substr($page_template, $startPos, $sectionLen); return trim($out); } /** * Function Save * * * @return string */ function Save() { global $page_template, $error_response, $session_dir; $tagSearch = array(); $varReplace = array(); $pattern = "RANDOM PICKUP LINE"; $errorMessage = ''; #$error_response = "No AIML category found. This is a Default Response."; $conversation_lines = '1'; $remember_up_to = '10'; $_SESSION['errorMessage'] = ''; // First off, write the config file $myPostVars = filter_input_array(INPUT_POST, FILTER_SANITIZE_STRING); ksort($myPostVars); $configContents = file_get_contents(_INSTALL_PATH_ . 'config.template.php'); $configContents = str_replace('[session_dir]', $session_dir, $configContents); if (!file_exists(_SESSION_PATH_)) { // Create the sessions folder, and set permissions mkdir(_SESSION_PATH_, 0755); // Place an empty index file in the sessions folder to prevent direct access to the folder from a web browser file_put_contents(_SESSION_PATH_ . 'index.html', ''); } foreach ($myPostVars as $key => $value) { $tagSearch[] = "[$key]"; $varReplace[] = $value; } $configContents = str_replace($tagSearch, $varReplace, $configContents); $saveFile = file_put_contents(_CONF_PATH_ . 'global_config.php', $configContents); // Now, update the data to the database, starting with making sure the tables are installed $dbh = $myPostVars['dbh']; $dbn = $myPostVars['dbn']; $dbu = $myPostVars['dbu']; $dbp = $myPostVars['dbp']; try { $dbConn = new PDO("mysql:host=$dbh;dbname=$dbn;charset=utf8", $dbu, $dbp); $dbConn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION); $dbConn->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC); } catch (Exception $e) { header('Content-type: text/plain'); var_dump($e); exit('Cannot connect to the database! ' . $e->getMessage()); } $sql = "show tables;"; $sth = $dbConn->prepare($sql); $sth->execute(); $row = $sth->fetch(); if (empty ($row)) { $sql = file_get_contents('new.sql'); $sth = $dbConn->prepare($sql); $sth->execute(); $affectedRows = $sth->rowCount(); } else { // Let's make sure that the srai lookup table exists try { $sql = 'select bot_id from srai_lookup;'; $sth = $dbConn->prepare($sql); $sth->execute(); $result = $sth->fetchAll(); } catch(Exception $e) { try { $sql = "DROP TABLE IF EXISTS `srai_lookup`; CREATE TABLE IF NOT EXISTS `srai_lookup` (`id` int(11) NOT NULL AUTO_INCREMENT, `bot_id` int(11) NOT NULL, `pattern` text NOT NULL, `template_id` int(11) NOT NULL, PRIMARY KEY (`id`), KEY `pattern` (`pattern`(64)) COMMENT 'Search against this for performance boost') ENGINE=MyISAM DEFAULT CHARSET=utf8 COMMENT='Contains previously stored SRAI calls' AUTO_INCREMENT=1;"; $sth = $dbConn->prepare($sql); $sth->execute(); } catch(Exception $e) { $errorMessage .= 'Could not add SRAI lookup table! Error is: ' . $e->getMessage(); } } } $sql = 'select `error_response` from `bots` where 1 limit 1'; $sth = $dbConn->prepare($sql); $sth->execute(); $row = $sth->fetch(); $error_response = $row['error_response']; $sql = 'select `bot_id` from `bots`;'; $sth = $dbConn->prepare($sql); $sth->execute(); $result = $sth->fetchAll(); if (count($result) == 0) { $sql_template = " INSERT IGNORE INTO `bots` (`bot_id`, `bot_name`, `bot_desc`, `bot_active`, `bot_parent_id`, `format`, `save_state`, `conversation_lines`, `remember_up_to`, `debugemail`, `debugshow`, `debugmode`, `error_response`, `default_aiml_pattern`) VALUES ([default_bot_id], '[bot_name]', '[bot_desc]', '[bot_active]', '[bot_parent_id]', '[format]', '[save_state]', '$conversation_lines', '$remember_up_to', '[debugemail]', '[debugshow]', '[debugmode]', '$error_response', '$pattern');"; $bot_id = 1; $sql = str_replace('[default_bot_id]', $bot_id, $sql_template); $sql = str_replace('[bot_name]', $myPostVars['bot_name'], $sql); $sql = str_replace('[bot_desc]', $myPostVars['bot_desc'], $sql); $sql = str_replace('[bot_active]', $myPostVars['bot_active'], $sql); $sql = str_replace('[bot_parent_id]', 1, $sql); $sql = str_replace('[format]', $myPostVars['format'], $sql); // "Use PHP from DB setting // "Update PHP in DB setting $sql = str_replace('[save_state]', $myPostVars['save_state'], $sql); $sql = str_replace('[conversation_lines]', $conversation_lines, $sql); $sql = str_replace('[remember_up_to]', $remember_up_to, $sql); $sql = str_replace('[debugemail]', $myPostVars['debugemail'], $sql); $sql = str_replace('[debugshow]', $myPostVars['debug_level'], $sql); $sql = str_replace('[debugmode]', $myPostVars['debug_mode'], $sql); $sql = str_replace('[error_response]', $error_response, $sql); $sql = str_replace('[aiml_pattern]', $pattern, $sql); try { $sth = $dbConn->prepare($sql); $sth->execute(); $affectedRows = $sth->rowCount(); $errorMessage .= ($affectedRows > 0) ? '' : ' Could not create new bot!'; } catch(Exception $e) { $errorMessage .= $e->getMessage(); } } $cur_ip = $_SERVER['REMOTE_ADDR']; $encrypted_adm_dbp = md5($myPostVars['adm_dbp']); $adm_dbu = $myPostVars['adm_dbu']; $sql = "select id from `myprogramo` where `user_name` = '$adm_dbu' and `password` = '$encrypted_adm_dbp';"; $sth = $dbConn->prepare($sql); $sth->execute(); $result = $sth->fetchAll(); if (count($result) == 0) { $sql = "insert ignore into `myprogramo` (`id`, `user_name`, `password`, `last_ip`) values(null, '$adm_dbu', '$encrypted_adm_dbp', '$cur_ip');"; try { $sth = $dbConn->prepare($sql); $sth->execute(); $affectedRows = $sth->rowCount(); $errorMessage .= ($affectedRows > 0) ? '' : ' Could not create new Admin!'; } catch(Exception $e) { $errorMessage .= $e->getMessage(); } } if (empty($errorMessage)) { $out = getSection('InstallComplete', $page_template); } else $out = getSection('InstallError', $page_template); return $out . $errorMessage; } `

**Comment by Dave-Morton** [comment]
You've posted a lot of information, but I don't see anywhere what version of PHP is installed on your system. I __did__ notice, however, that this was part of your post:

>We're sorry, but Program O requires PHP version 5.3.0 or greater to function. Please ask your hosting provider to upgrade.

This would indicate that you're using a version of PHP that is 5.2 or lower. If that's the case, you'll never get Program O to work, because some of the functions and features of Program O require PHP 5.3 or higher to work. Additionally, it seems that PDO support is not enabled in your system, which is also required. Once we get the PHP version problem addressed, enabling PDO support is moderately simple. Then we can explore further.

**Comment by pokephreak** [comment]
Thanks Dave,

php -v shows that version 5.5.9-1ubuntu4.20 is installed. Which is one of the reasons I don't trust any errors I see in the output. The output also contains portions of the php, making me think it's being rendered as text instead of being run as php.

This is a remote server at hostinger.co.uk

**Comment by Dave-Morton** [comment]
I noticed that, too, which leads me to wonder if the PHP engine might be experiencing some sort of glitch. I have a VM with a very similar setup, and haven't seen this type of behaviour before. Let me do some checking, and I'll get back on this later tonight or tomorrow.

**Comment by pokephreak** [comment]
Thanks very much Dave, much appreciated.

My [phpinfo](php.net/manual/en/function.phpinfo.php) can be found here http://31.220.59.44/info.php which only confuses me more - I don't understand how this can work at /var/www/html/info.php and program-o doesn't, but is also found at /var/www/html/Program-O/ .....

Thanks again for your time Dave.

**Comment by Dave-Morton** [comment]
I followed the link to your PHPINFO, then deleted the link for security and privacy reasons (something of that sort would be best sent via email, BTW). I'll look over it and do some testing on the installation now that I have the IP address, then I'll let you know what I find.

**Comment by Dave-Morton** [comment]
@pokephreak can you send me an email regarding this, as well? there are some questions that I want to ask, the answers to which should not be made public. My email address is dmorton(at)geekcavecreations(dot)com

**Comment by pokephreak** [comment]
Thanks a lot. It will all be locked down once I figure out what I'm doing wrong with program-o.

**Comment by pokephreak** [comment]
Sent you an email Dave.

**Comment by Dave-Morton** [comment]
Got the email and sent a reply. Thanks. :)

**Comment by Dave-Morton** [comment]
Ok, it appears that we _may_ have found the issue. @pokephreak please list for everyone both the browser you were using when you saw the errors, and all of the addons/extensions for that browser. This may help others in the future who have similar issues. Thanks,

**Comment by pokephreak** [comment]
Using Firefox 47.0 for Ubuntu, I was not able resolve the issue. However, a good clue is that everything works fine in chrome. So server side, everything looks good. If you're having a similar issue, the problem is probably with your browser. I have a lot of development plugins, extensions and some privacy modifications. 

To find the culprit, either disable all the extensions/addons and, assuming this results in program-o working, start switching them back on in order until you find the one that breaks the installation script. Alternatively disable them in order until the install script works, then you've found the extension that breaks the install script.

So to summarize, the issue was a Firefox modification, program-o works and can be accessed with Chrome.

Thanks a lot for taking the time to help Dave.
